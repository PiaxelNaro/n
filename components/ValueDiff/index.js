// @flow
import React from 'react';
import styled from 'styled-components';

const COLORS = {
    positive: '#008c23',
    negative: '#990000',
    neutral: '#ccc',
};

export const Wrapper = styled.span`
    color: ${props => COLORS[props.type]};
    font-family: Helvetica, Arial, sans-serif;
`;

const Unit = styled.span`
    font-size: 10px;
`;

const ValueDiff = ({
    value,
    type,
    unit = '',
}: {
    value: number,
    type?: 'value' | 'growth',
    unit?: string,
}) => {
    if (value === 0) {
        return <Wrapper type="neutral">=</Wrapper>;
    }

    let glyph;
    if (type === 'growth') {
        glyph = value > 0 ? '+' : '-';
    } else if (type === 'value') {
        glyph = value > 0 ? '▲' : '▼';
    }

    return (
        <Wrapper type={value > 0 ? 'positive' : 'negative'}>
            <Unit>{glyph}</Unit>
            {Math.abs(value)}
            <Unit>{unit}</Unit>
        </Wrapper>
    );
};

export default ValueDiff;
