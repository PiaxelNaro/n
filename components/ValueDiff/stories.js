import React from 'react';
import { storiesOf } from '@kadira/storybook';
import ValueDiff from '.';
import Table from '../Table';

const formatValueDiff = value => <ValueDiff value={value} type="value" />;
const formatMoneyDiff = value => <ValueDiff value={value} type="value" unit="$" />;
const formatPercDiff = value => <ValueDiff value={value} type="growth" unit="%" />;

storiesOf('ValueDiff', module).add('Types and unit', () => (
    <Table
        columns={[
            {
                key: 'value',
                label: 'Value',
            },
            {
                key: 'perc',
                label: 'Percentage',
            },
            {
                key: 'money',
                label: 'Money',
            },
        ]}
        items={[
            {
                value: 0.2678,
                money: 0.2457,
                perc: 0.2234,
            },
            {
                value: -0.5457,
                money: -5.235,
                perc: -0.1567,
            },
            {
                value: 0,
                money: 0,
                perc: 0,
            },
        ]}
        formatters={{
            value: formatValueDiff,
            money: formatMoneyDiff,
            perc: formatPercDiff,
        }}
    />
));
