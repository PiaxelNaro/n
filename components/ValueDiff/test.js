// @flow
import React from 'react';
import { shallow } from 'enzyme';
import ValueDiff, { Wrapper } from '.';

describe('<ValueDiff />', () => {
    it('renders value with unit', () => {
        expect(
            shallow(<ValueDiff value={0.234551} type="value" unit="%" />)
        ).toMatchSnapshot();
        expect(
            shallow(<ValueDiff value={0.234551} type="growth" unit="%" />)
        ).toMatchSnapshot();
    });

    it('renders value without unit', () => {
        expect(shallow(<ValueDiff value={-3.2} type="value" />)).toMatchSnapshot();
        expect(shallow(<ValueDiff value={-3.2} type="growth" />)).toMatchSnapshot();
    });

    it('renders neutral value', () => {
        expect(shallow(<ValueDiff value={0} type="value" />)).toMatchSnapshot();
        expect(shallow(<ValueDiff value={0} type="growth" />)).toMatchSnapshot();
    });

    it('renders value without type', () => {
        expect(shallow(<ValueDiff value={1} />)).toMatchSnapshot();
        expect(shallow(<ValueDiff value={-1} />)).toMatchSnapshot();
    });

    it('sets color properly', () => {
        expect(shallow(<Wrapper type="positive" />)).toMatchSnapshot();
    });
});
