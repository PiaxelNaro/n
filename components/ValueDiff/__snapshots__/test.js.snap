// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ValueDiff /> renders neutral value 1`] = `
<ValueDiff__Wrapper
  type="neutral"
>
  =
</ValueDiff__Wrapper>
`;

exports[`<ValueDiff /> renders neutral value 2`] = `
<ValueDiff__Wrapper
  type="neutral"
>
  =
</ValueDiff__Wrapper>
`;

exports[`<ValueDiff /> renders value with unit 1`] = `
<ValueDiff__Wrapper
  type="positive"
>
  <ValueDiff__Unit>
    ▲
  </ValueDiff__Unit>
  0.234551
  <ValueDiff__Unit>
    %
  </ValueDiff__Unit>
</ValueDiff__Wrapper>
`;

exports[`<ValueDiff /> renders value with unit 2`] = `
<ValueDiff__Wrapper
  type="positive"
>
  <ValueDiff__Unit>
    +
  </ValueDiff__Unit>
  0.234551
  <ValueDiff__Unit>
    %
  </ValueDiff__Unit>
</ValueDiff__Wrapper>
`;

exports[`<ValueDiff /> renders value without type 1`] = `
<ValueDiff__Wrapper
  type="positive"
>
  <ValueDiff__Unit />
  1
  <ValueDiff__Unit />
</ValueDiff__Wrapper>
`;

exports[`<ValueDiff /> renders value without type 2`] = `
<ValueDiff__Wrapper
  type="negative"
>
  <ValueDiff__Unit />
  1
  <ValueDiff__Unit />
</ValueDiff__Wrapper>
`;

exports[`<ValueDiff /> renders value without unit 1`] = `
<ValueDiff__Wrapper
  type="negative"
>
  <ValueDiff__Unit>
    ▼
  </ValueDiff__Unit>
  3.2
  <ValueDiff__Unit />
</ValueDiff__Wrapper>
`;

exports[`<ValueDiff /> renders value without unit 2`] = `
<ValueDiff__Wrapper
  type="negative"
>
  <ValueDiff__Unit>
    -
  </ValueDiff__Unit>
  3.2
  <ValueDiff__Unit />
</ValueDiff__Wrapper>
`;

exports[`<ValueDiff /> sets color properly 1`] = `
<span
  className="ValueDiff__Wrapper-s1lggxkb-0 bKlvlc"
  type="positive"
/>
`;
